vertical-pod-autoscaler/enhancements/7397-max-allowed-recommendation/README.md

```diff
@@ -0,0 +1,65 @@
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+   - [Goals](#goals)
+- [Proposal](#proposal)
+- [Design Details](#design-details)
+   - [Proposed Solution](#proposed-solution)
+   - [Alternatives](#alternatives)
+   - [Challenges](#challenges)
+   - [Test Plan](#test-plan)
+<!-- /toc -->
+
+## Summary
+VPA currently allows configuring the maximum allowed recommendation per resource using the field `.spec.resourcePolicy.containerPolicies[].maxAllowed`. This helps ensure that recommendations don't exceed the allocatable capacity of the largest node. Without this limit, a high recommendation could make the Pod unschedulable, causing downtime. We aim to improve the VPA recommender by making it aware of the maximum node size, preventing it from generating recommendations that would render Pods unschedulable.
+
+## Motivation
+In dynamic environments, manual configuration of .maxAllowed can be error-prone, especially as cluster sizes and node configurations change over time. Automating this process by making the VPA recommender aware of the largest node size ensures more reliable scaling decisions. This helps prevents situations where Pods become unschedulable due to excessive resource recommendations.
+
+### Goals
+* Make Recommendations More Accurate and Easier to Schedule: Try to have the VPA recommender suggest resources that fit within the largest node's capacity, helping pods get scheduled more easily and lowering the chance of downtime.
+* Adapt to Changing Environments: Adjust recommendations as the cluster size or node configurations change to better reflect the current state of the cluster.
+
+## Proposal
+We propose to improve the VPA recommender by making it aware of the maximum node size, preventing it from generating recommendations that would render Pods unschedulable. This will involve querying for cluster information and adjusting recommendations accordingly.
+
+## Design Details
+
+### Proposed Solution
+We propose to use the Cluster Autoscaler's provisioning-request API to obtain information about the maximum node size. This approach aligns well with CA's design and capabilities. The implementation will include:
+
+1. Feature Flag:
+   The behavior of querying Cluster Autoscaler will be controlled by a clearly defined flag. This ensures that users can explicitly enable or disable this feature, providing clarity on whether they can rely on CA for recommendation limits.
+
+2. Caching Mechanism:
+   To optimize performance and reduce the load on the Cluster Autoscaler, a caching mechanism will be implemented. This would involve storing the retrieved node size information for a configurable period before refreshing.
+
+3. Cluster Autoscaler Availability Warning:
+   The system should implement a warning mechanism to alert users when Cluster Autoscaler is not present in the cluster. This warning will help users understand why the feature might not be functioning as expected and guide them towards proper configuration.
```

Comment:
How would that warning look like? A status condition on VPA object? Something else? I'd expect a KEP to describe in detail any user-facing API.

Reply:
Understood. I considered logging those warnings and emitting them as events, but using a status condition on the VPA object sounds like a better approach. Ill adjust.